mapper/software_management/README.md

```diff
@@ -0,0 +1,110 @@
+## Components
+
+### Thin-Edge-Json
+
+Responsibilities:
+* Define the API between a cloud provider and the DM-agent
+* Ability to convey as well:
+   * desired software lists
+   * software operation lists
+   * device profile
+   * current software lists
+   * operation status
+* Link a response to a request
+
+Out of scope:
+* Send the operation logs to the cloud.
+
+Open questions:
+* How to shorten the current software list sent by a device?
+
+### Specific mapper - e.g. C8y Mapper
+
+Responsibilities:
+* Register the device, so it will receive the software management requests
+* Listen and translate the operations received from the cloud.
+* Forward these operations to the DM-agent and listen for the responses.
+* Translate the DM-Agent responses and forward them to the cloud.
+
+Open questions:
+* Each software update message has an id, to be used for the response ...
+   * Do we have to report the result in one go ... or can we have several partial responses using the same id?
+
+
+### DM-Agent
+
+Responsibilities:
+* Listen the mappers for software management operations.
+* Translate each software management request into a sequence of software management operations.
+* Use a database to store the schedule operations and their status.
+* Use the Plugin store to find the appropriate plugin for each operation.
+* Apply the operation schedule, updating the database accordingly.
+
+Out of scope:
+* Manage the dependencies
+
+Open questions:
+* How to translate a *desired software list* into a sequence of operations?
+ 
+
+### DM-Agent database
+
+Responsibilities:
+* Store the operations to schedule
+* Store the result of each operation
+  * One log file per install / uninstall command
+
+Open questions:
+* How to manage process start / stop / restart *while* applying a list of updates.
+* Is the `dm-agent` a mapper?
```

Comment:
I see a mapper as a process reading messages from input topics and sending a translated version to some output topic (or an error message to an error topic).

* A translation is supposed to be instantaneous and never in some scheduled state where the request has been sent but no response received yet. One can no more assume that the mapper will not been interrupted after a response published and before the input acknowledged. One needs to persist that a request has been sent and to resume any scheduled requests on restart.
* The strange requirements about the reverse order of the responses, is yet another issue that it would be difficult to generalize under a single mapper concept.
* Furthermore, the translation of a cloud specific software-management message into a series of software-operation might depends on the current system state.